src/Arara/Process/IPC/Semaphore.php

```diff
@@ -0,0 +1,132 @@
+<?php
+
+namespace Arara\Process\IPC ;
+
+/**
+ * This class provides Interprocess communication using UNIX Semaphores
+ *
+ * Code written by Nikolay and copied into Arara/Process under the terms of
+ * the MIT license.
+ *
+ * Full credit to Nikolay for his excellent class
+ * 
+ * @license http://opensource.org/licenses/MIT
+ * @author Nikolay Bondarenko http://misterion.ru 
+ */
+class Semaphore
+{
+    /**
+     * @var resource
+     */
+    protected $mutex;
+
+    /**
+     * @var string
+     */
+    protected $file;
+
+    /**
+     * @var bool
+     */
+    protected $isAcquired;
+
+    /**
+     * Creator process pid
+     *
+     * @var int
+     */
+    protected $pid;
+
+    public function __construct($key = null)
+    {
+        $this->pid = getmypid();
+
+        $this->isAcquired = false;
+        if ($key) {
+            $semKey = $key;
+        } else {
+            $this->file = tempnam(sys_get_temp_dir(), 's');
+            $semKey = ftok($this->file, 'a');
+        }
+
+        $this->mutex = sem_get($semKey, 1); //auto_release = 1 by default
+        if (!$this->mutex) {
+            throw new \RuntimeException('Unable to create the semaphore');
+        }
+    }
+
+    public function __destruct()
+    {
+        if ($this->isAcquired()) {
+            $this->release();
+        }
+
+        if ($this->pid === getmypid()) {
+            $this->remove();
+        }
```

Comment:
I don't think this approach is wrong. We must take care into the two approaches.

If remove automatically by using PID as reference
- :+1:: You don't need to remove it manually
- :-1:: You must instantiate the class before instantiate any child and pass the object as reference
- :-1:: You cannot use the semaphore unless you are spawning, otherwise you will remove it in some process but other process will use it

If this decision is up to you
- :-1:: You need to do it manually
- :+1:: You don't need to spawn to use the semaphore
- :+1:: You can use the same semaphore on process which have no relation